aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -61,6 +83,45 @@ static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, b
   return r;
 }
 
+template <typename T>
+static inline bool is_null_or_equal_to(const c10::optional<T>& test, const T& value) {
+  if (!test.has_value()) {
+    return true;
+  }
+  return test.value() == value;
+}
+
+static inline Tensor to_impl(
+    const Tensor& self,
+    c10::optional<ScalarType> dtype,
+    c10::optional<Layout> layout,
+    c10::optional<Device> device,
+    c10::optional<bool> pin_memory,
+    bool non_blocking,
+    bool copy,
+    c10::optional<c10::MemoryFormat> optional_memory_format) {
+  auto memory_format = optional_memory_format.value_or(MemoryFormat::Preserve);
+
+  // fast path
+  if (is_null_or_equal_to(dtype, self.dtype().toScalarType()) &&
+      is_null_or_equal_to(layout, self.layout()) &&
+      is_null_or_equal_to(device, self.device()) &&
+      !copy &&
+      // TODO(#62027): The following seems to be wrong.
+      // self.suggest_memory_format() suggests MemoryFormat::Contiguous
+      // for non-contiguous tensors that don't look ChannelsLast.
+      // This leads to `x.to(memory_format=torch.contiguous_format)`
+      // not always returning a contiguous tensor. Maybe the condition should
+      // actually be `self.is_contiguous(memory_format)`?
+      (memory_format == MemoryFormat::Preserve ||
+       self.suggest_memory_format() == memory_format)) {
+    return self;
+  }
+
+  return at::_to_copy(
```

Comment:
does it increase the number of dispatches that `to` goes through? Before it used to be `to` and `copy_`, now it's `to`, `_to_copy` and `copy_`?

Reply:
Yes, that is correct. There are benchmarks in the PR body that demonstrate this. If we care about overhead, then in the future we can make `_to_copy` directly call TensorIterator instead of calling `copy_`.